spec/registry.md

```diff
@@ -0,0 +1,156 @@
+* Forbid edits to certain function options (e.g. currency options).
+* Autocomplete function and option names.
+* Display on-hover tooltips for function signatures with documentation.
+* Display/edit known message metadata.
+* Restrict input in GUI by providing a dropdown with all viable option values.
+
+## Data Model
+
+The registry contains descriptions of function signatures.
+[`registry.dtd`](./registry.dtd) describes its data model.
+
+The main building block of the registry is the `<function>` element.
+It represents an implementation of a custom function available to translation at runtime.
+A function defines a human-readable _description_ of its behavior
+and one or more machine-readable _signatures_ of how to call it.
+Named `<pattern>` elements can optionally define regex validation rules for input, option values, and variant keys.
+
+A _signature_ defines one particular set of arguments and named options that can be used together in a single call to the function.
+`<formatSignature>` corresponds to a function call inside a placeholder inside translatable text.
+`<matchSignature>` coresponds to a function call inside a selector.
+Signatures with a non-empty `locales` attribute are locale-specific and only available in translations in the given languages.
+
+A signature may define the positional argument of the function with the `<input>` element.
+A signature may also define one or more `<option>` elements representing _named options_ to the function.
+Options are optional by default,
+unless the `required` attribute is present.
+They accept either a finite enumeration of values (the `values` attribute)
+or validate they input with a regular expression (the `pattern` attribute).
+Read-only options (the `readonly` attribute) can be displayed to translators in CAT tools, but may not be edited.
+
+Matching-function signatures additionally include one or more `<match>` elements
+to define the keys against which they can match when used as selectors.
+
+## Example
+
+The following `registry.xml` is an example of a registry file
+which may be provided by an implementation to describe its built-in functions.
+For the sake of brevity, only `locales="en"` is considered.
+
+```xml
+<?xml version="1.0" encoding="UTF-8" ?>
+<!DOCTYPE registry SYSTEM "./registry.dtd">
+
+<registry>
+    <function name="platform">
+        <description>Match the current OS.</description>
+        <matchSignature>
+            <match values="windows linux macos android ios"/>
+        </matchSignature>
+    </function>
+
+    <pattern id="anyNumber" regex="-?[0-9]+(\.[0-9]+)"/>
+    <pattern id="positiveInteger" regex="[0-9]+"/>
+    <pattern id="currencyCode" regex="[A-Z]{3}"/>
+
+    <function name="number">
+        <description>
+            Format a number.
+            Match a numerical value against CLDR plural categories or against a number literal.
+        </description>
+
+        <matchSignature locales="en">
```

Comment:
I'm very curious about the `locales` attribute--curious in this case is a euphemism for "nervous about". I would prefer if functions were mostly locale neutral (I can declare, for example, a number format in any message in any locale).

Reply:
Oh, I think I like it!

Reply:
Filed #410 to discuss this further.